Migrate CardinalityIT test to AggregatorTestCase

salvatore-campagna opened the pull request:
Replace integration tests with unit tests for cardinality aggregations.
Unit tests are generally faster and easier to test against.

This is part of the transition from integration to unit tests described by #42893

salvatore-campagna committed:
Fix spelling mistake

salvatore-campagna committed:
Extract assertCacheHits and assertCacheMisses methods

salvatore-campagna committed:
Remove unused hash parameter

salvatore-campagna committed:
Migrate tests from CardinalityIT to CardinalityAggregatorTests

Replace integration tests with unit tests for cardinality aggregations.
Unit tests are generally faster and easier to test against.

salvatore-campagna committed:
Remove non-required changes for two tests

csoulios reviewed (commented):
Very good first effort. Perhaps you should override the `getSupportedValuesSourceTypes()` method so that we test the matrix of supported/unsupported field types.

Also, I think that some aggregation tests from `CardinalityIT` have not been ported to `CardinalityAggregatorTests` (`testAsSubAgg`, `testScriptCaching`, `testPartiallyUnmapped` and the "value script" tests) (check `MaxAggregatorTests` on how to implement some of them)

salvatore-campagna committed:
Test for supported and unsupported field types

Overriding getSupportedValuesSourceTypes and createAggBuilderForTypeTest
results in testing all the supported value types for the cardinality
aggregator. The cardinality aggregator supports all core values types.

salvatore-campagna committed:
Add a cardinality sub-aggregation test

This test uses a terms aggregation as a top level aggregation
and a cardinality aggregation as a sub-aggretation.

salvatore-campagna committed:
Add tests using a '_value' script for single and multi-valued fields

salvatore-campagna committed:
Test cardinality aggregator with partially unmapped index

salvatore-campagna committed:
Fix code format violations

salvatore-campagna committed:
Add a unit test checking aggregation caching

salvatore-campagna committed:
Merge branch 'master' into feature/migrate-cardinalityit-to-aggregator-test-case

salvatore-campagna committed:
Merge branch 'master' into feature/migrate-cardinalityit-to-aggregator-test-case

salvatore-campagna commented:
@elasticsearchmachine elasticsearch-ci/rest-compatibility

salvatore-campagna commented:
@elasticmachine rest-compatibility

salvatore-campagna commented:
@elasticmachine elasticsearch-ci/rest-compatibility

salvatore-campagna commented:
@elasticmachine run rest-compatibility

salvatore-campagna commented:
@elasticmachine run elasticsearch-ci/rest-compatibility

csoulios commented on server/src/test/java/org/elasticsearch/search/aggregations/metrics/CardinalityAggregatorTests.java:
```diff
@@ -112,6 +213,180 @@ public void testQueryFiltersAll() throws IOException {
+        }, mappedFieldTypes);
+    }
+
+    public void testMultiValuedStringVaalueScript() throws IOException {
```
nit: there's a small typo here

s/testMultiValuedStringVaalueScript/testMultiValuedStringValueScript

csoulios commented on server/src/test/java/org/elasticsearch/search/aggregations/metrics/CardinalityAggregatorTests.java:
```diff
@@ -138,6 +413,175 @@ public void testUnmappedMissingNumber() throws IOException {
+            );
+            iw.addDocument(Arrays.asList(new SortedNumericDocValuesField("numbers", 11), new SortedNumericDocValuesField("numbers", 12)));
+            iw.addDocument(Arrays.asList(new SortedNumericDocValuesField("numbers", 12), new SortedNumericDocValuesField("numbers", 13)));
+            iw.addDocument(
+                Arrays.asList(new SortedNumericDocValuesField("numbers", 12), new SortedNumericDocValuesField("numbers", 13))
+
+            );
```
nit: Not sure why this is formatted different than the lines above.

csoulios commented on server/src/test/java/org/elasticsearch/search/aggregations/metrics/CardinalityAggregatorTests.java:
```diff
@@ -152,6 +596,129 @@ public void testUnmappedMissingGeoPoint() throws IOException {
+        unmappedDirectory.close();
+    }
+
+    private static byte[] hash(final byte[] value) {
```
This method is not needed. Please remove that too

csoulios reviewed (commented):
Add couple more comments.

nik9000 commented on server/src/test/java/org/elasticsearch/search/aggregations/metrics/CardinalityAggregatorTests.java:
```diff
@@ -112,6 +213,180 @@ public void testQueryFiltersAll() throws IOException {
+                )
+            );
+            iw.addDocument(
+                Arrays.asList(
```
It's not important, but in more modern code we've moved to `List.of` here. The list is makes is immutable which is nice. It doesn't make a difference here, but it's still nice.

nik9000 commented on server/src/test/java/org/elasticsearch/search/aggregations/metrics/CardinalityAggregatorTests.java:
```diff
@@ -152,6 +596,129 @@ public void testUnmappedMissingGeoPoint() throws IOException {
+
+        final Directory directory = newDirectory();
+        final RandomIndexWriter indexWriter = new RandomIndexWriter(random(), directory);
+        final int numDocs = 10;
```
One of the `testCase` style methods in the superclass ought to be able to do this stuff for you. It's nice to use them mostly because they add extra randomization to the reduction and the circuit breaker.

salvatore-campagna committed:
Remove test using hashing

salvatore-campagna committed:
Replace Arrays.asList with List.of

salvatore-campagna committed:
Refactor test to reuse testCase from parent class

salvatore-campagna committed:
Fix a typo and code format violations

salvatore-campagna committed:
Refactor test to reuse testCase for global aggregation too

salvatore-campagna commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into feature/migrate-cardinalityit-to-aggregator-test-case

nik9000 reviewed (approved):
LGTM

csoulios reviewed (approved):
LGTM

salvatore-campagna commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into feature/migrate-cardinalityit-to-aggregator-test-case

csoulios commented:
Congrats on your first PR getting merged! :)
